### PR TITLE
Xml proposal implementation: XML node structure change

### PR DESCRIPTION
**Pull request by rdhananjaya**
## Purpose
This PR adds the new XML node structure, xml parser, xml serializer as a part of revamping xml support. This is a part of https://github.com/ballerina-platform/ballerina-lang/issues/19570

Fixes https://github.com/ballerina-platform/ballerina-lang/issues/19832

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    @Override
+    public void close() throws IOException {
+        try {
+            xmlStreamWriter.close();
+        } catch (XMLStreamException e) {
+            throw new BallerinaException(e);
+        }
+    }
+
+    public void write(BXML xmlValue) {
```
Comment (dulvinw) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java]: What if we define `throws` since this a public method?

Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java]: It doesn't throw any checked exceptions, in that case, do we need to do that.

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+        if (!elementNSUsageFoundInAttribute && !qName.getNamespaceURI().isEmpty()) {
+            xmlStreamWriter.setDefaultNamespace(qName.getNamespaceURI());
+            return qName.getNamespaceURI();
+        }
+
+        // Undeclare default namespace for this element, if outer elements have redefined default ns and this
+        // element doesn't have NS URI in it's name.
+        if ((qName.getNamespaceURI() == null || qName.getNamespaceURI().isEmpty())) {
+            for (String s : currentNSLevel) {
+                if (s.startsWith("xmlns")) {
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java]: Shall we use constants for "xmlns", "", "}", etc.?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/XMLTreeBuilder.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+import static javax.xml.stream.XMLStreamConstants.DTD;
+import static javax.xml.stream.XMLStreamConstants.END_DOCUMENT;
+import static javax.xml.stream.XMLStreamConstants.END_ELEMENT;
+import static javax.xml.stream.XMLStreamConstants.PROCESSING_INSTRUCTION;
+import static javax.xml.stream.XMLStreamConstants.START_ELEMENT;
+
+/**
+ * XML tree builder for Ballerina xml node structure using {@code XMLStreamReader}.
+ *
+ * @since 1.1.0
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/XMLTreeBuilder.java]: Versions need to be updated?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/XMLTreeBuilder.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+        try {
+            xmlStreamReader = xmlInputFactory.createXMLStreamReader(stringReader);
+        } catch (XMLStreamException e) {
+            handleXMLStreamException(e);
+        }
+    }
+
+    private void handleXMLStreamException(Exception e) {
+        // todo: do e.getMessage contain all the information? verify
+        throw new BallerinaException(e.getMessage(), e);
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/XMLTreeBuilder.java]: I guess the reason should be `{ballerina/lang.xml}XxxError`

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    private Deque<Set<String>> parentNSSet;
+    private int nsNumber;
+
+
+    public BallerinaXMLSerializer(OutputStream outputStream) {
+        try {
+            xmlStreamWriter = xmlOutputFactory.createXMLStreamWriter(outputStream);
+            parentNSSet = new ArrayDeque<>();
+        } catch (XMLStreamException e) {
+            throw new BallerinaException(e);
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java]: Error reasons need to be fixed?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        }
+    }
+
+    private void writeXMLPI(XMLPi xmlValue) throws XMLStreamException {
+        xmlStreamWriter.writeProcessingInstruction(xmlValue.getTarget(), xmlValue.getData());
+    }
+
+    private void writeXMLComment(XMLComment xmlValue) throws XMLStreamException {
+        xmlStreamWriter.writeComment(xmlValue.getTextValue());
+
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/BallerinaXMLSerializer.java]: Extra line?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/XMLNonElementItem.java [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    }
+
+    @Override
+    @Deprecated
+    public void removeChildren(String qname) {
+
+    }
+
+    @Override
+    public abstract OMNode value();
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/XMLNonElementItem.java]: Do we need this?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/XMLText.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    }
+
+    @Override
+    public XMLNodeType getNodeType() {
+        return XMLNodeType.TEXT;
+    }
+
+    @Override
+    public OMNode value() {
+            CharacterDataImpl characterData = new CharacterDataImpl();
```
Comment (MaryamZi) [inline bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/XMLText.java]: formatting issue?

File: bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/XMLNonElementItem.java (hunk 7 above)
Comment (rdhananjaya) [reply bvm/ballerina-runtime/src/main/java/org/ballerinalang/jvm/values/XMLNonElementItem.java]: This is still being used for testing (Jvm values to BVM values conversion)

**Comment by rdhananjaya** [comment]
This PR was mistakenly merged by pushing to `ballerina-platform` instead of 'rdhananjaya`.
This PR is replaced by https://github.com/ballerina-platform/ballerina-lang/pull/20774

**Pull request merged**